src/MDOP.jl

```diff
@@ -87,3 +220,134 @@ function beta_statistic(Y::Dataset, s::Vector; τs = 0:50 , w::Int = 1)
     return vec(β)
 end
+
+
+"""
+    embed2(Y::Dataset,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+embedding vector `Y`.
+"""
+function embed2(Y::Dataset,s::Vector{Float64},τ::Int)
+    N = size(Y,1)   # length of input trajectory
+    NN = size(Y,2)  # dimensionality of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,NN+1)
+    # fill vector up until index M
+    @inbounds for i = 1:NN
+        Y_new[:,i] = Y[1:M,i]
+    end
+    # add lagged component of s
+    Y_new[:,NN+1] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
+
+"""
+    embed2(y::Vector,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+time series `y`.
+"""
+function embed2(y::Vector{Float64},s::Vector{Float64},τ::Int)
+    N = size(y,1)   # length of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,2)
+    # fill vector up until index M
+    Y_new[:,1] = y[1:M]
+    # add lagged component of s
+    Y_new[:,2] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
+
+
+"""
+    fnn(NNdist,NNdistnew,r::Float64=2.0) -> FNNs
+fnn computes the amount of false nearest neighbors when adding another component
+to a given (vector-) time series. This new component is the `τ`-lagged version
+of a univariate time series. 'NNdist' is storing the distances of the nearest
+neighbor for all considered fiducial points and 'NNdistnew' is storing the
+distances of the nearest neighbor for each fiducial point in one embedding
+dimension higher using a given `τ`. The obligatory threshold `r` is by default
+set to 2.
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+"""
+function fnn(NNdist::T,NNdistnew::T;r::Float64=2.0) where {T}
+    @assert length(NNdist) == length(NNdistnew) "Both input vectors need to store the same number of distances."
+
+    # convert array of arrays into simple vectors, since we only look at K=1
+    NN_old = zeros(length(NNdist))
+    NN_new = zeros(length(NNdistnew))
+    for i = 1:length(NNdist)
+        NN_old[i]=NNdist[i][1]
+        NN_new[i]=NNdistnew[i][1]
+    end
+
+    # first ratio
+    ratio1 = 1/r    # since we input only z-standardized data with unit
+    # construct statistic
+    ratio2 = NN_new./NN_old
+
+    cond1 = ratio2 .> r
+    cond2 = NN_old .< ratio1
+
+    fnns = sum(cond1.*cond2)
+    fnns2 = sum(NN_old .< ratio1)
+
+    # store fraction of valid nearest neighbors
+    if fnns2 == 0
+        FNN = NaN;
+    else
+        FNN = fnns/fnns2
+    end
+
+    return FNN
+end
+
+
+"""
+    estimate_maximum_delay(s::Vector; tw=1:50, samplesize::Float=1.0)) -> τ_max, L
+computes an upper bound for the search of optimal delays, when using `MDOP`
```

Comment:
Typically documentation strings in Julia are stated in imperative second person: "Compute an upper bound..." instead of "computes"